docs/quarantine.md

```diff
@@ -102,3 +102,27 @@ After merging this PR the test will be out of quarantine.
 [2]: https://www.thoughtworks.com/en-us/insights/blog/no-more-flaky-tests-go-team
 [3]: https://docs.gitlab.com/ee/development/testing_guide/flaky_tests.html#quarantined-tests
 [on testgrid]: https://testgrid.k8s.io/kubevirt-periodics
+
+# Test Lane Quarantine
+
+There can be cases where required test lanes are flaking in a clustered manner with a large
+number of tests failing each time. This can cause major delays to merging important pull
+requests and overload CI. Individual test case quarantining does not make sense in this case.
+
+A required test lane should be made optional if the following criteria are met:
+* The required test lane has a failure rate higher than 25% in the last 24 hours
+* More than ten individual test cases are causing the required test lane to fail
+* The SIG responsible for the required test lane is unable to deliver a fix for the
+flake within 24 hours
```

Comment:
Delivering a fix for a one flake is sometimes a job for days not hours. I am curious how do we set some deadline.

Let's say we encounter clustered failures. What do we gain by  merging PRs/quarantine? 
Imho if we continue merging PRs with a blocking lane we risk that something new will fall through our gating, potentially extending the issue.  On the other hand, not merging PRs for few hours/days, assuming we have people urgently attending, we don't risk new failures being introduced. One obvious negative is to not be able to merge fixes for stable branches but here we can simply override. Other than that we still have an option to backtrack what PRs might have cause this and revert them.

So I am wondering if we are setting up our self for failure here.

On the other hand I believe we need to close the loop on the existing quarantine definition, where we don't make sure these tests are being worked on with priority. We continue releasing without potential valuable feedback from them.

Reply:
I think the current proposed text highlights that unstable production or test code should not slow down the whole project. If a specific SIG or subproject suffers from too much instability, the contributors as a whole should re-prioritize their efforts to solve the problem over anything else. One way to do it is by blocking any new changes to get in, solving exactly the cases mentioned in the arguments (e.g. letting bugs get in because it is not covered).

> Other than that we still have an option to backtrack what PRs might have cause this and revert them.

I think this is one of the best solutions for an immediate action.
For new tests or new production code that can be recognized as causing the problem, it is this or a real fix, whichever comes first.

> So I am wondering if we are setting up our self for failure here.

One can argue that it will just expose an already existing failure, making it visible and not hidden under the hood.

> On the other hand I believe we need to close the loop on the existing quarantine definition, where we don't make sure these tests are being worked on with priority. We continue releasing without potential valuable feedback from them.

I agree. I would expect the quarantined tests owners to be be responsible for this. If they do not have the time to maintain it, then the tests should be removed after a period of time. There is no choice but to balance between our hope to cover as much as possible and our ability to maintain all the instability of such tests (i.e. to fix them).

Reply:
> Delivering a fix for a one flake is sometimes a job for days not hours. I am curious how do we set some deadline.

Exactly if its going to take longer than a day to fix then we know that we shouldn't leave the lane in question required as it is going to block other SIGs.

> 
> Let's say we encounter clustered failures. What do we gain by merging PRs/quarantine? Imho if we continue merging PRs with a blocking lane we risk that something new will fall through our gating, potentially extending the issue. 

I would see the risk of breakage as an added incentive for the responsible SIG to make this a top priority as it should be.  The responsible SIG should really not be merging PRs related to the flaking test lane during this quarantine period which should lower the risk of breaking something. 

Again these scenarios are relatively rare but the aim of this procedure is to ensure that its all hand on deck from the responsible SIG to fix the broken lane.

>On the other hand, not merging PRs for few hours/days, assuming we have people urgently attending, we don't risk new failures being introduced. One obvious negative is to not be able to merge fixes for stable branches but here we can simply override. Other than that we still have an option to backtrack what PRs might have cause this and revert them.
> 
> So I am wondering if we are setting up our self for failure here.
> 
> On the other hand I believe we need to close the loop on the existing quarantine definition, where we don't make sure these tests are being worked on with priority. We continue releasing without potential valuable feedback from them.

Reply:
> Delivering a fix for a one flake is sometimes a job for days not hours. I am curious how do we set some deadline.

It may be hard to set a deadline, but we must have one. Let us assume that the windows lane is suddenly failing 90% of the times, making PRs merge after ~10 retries and many hours. @xpivarc for how long would you say we should accept this? Two days? A week? A month?

I recall having the fossa test fail 100% of the time. We attempted to fix it for a few days, failed to do so, so converted it to non-voting. Sure, this means that non-free code can sneak in; but the alternative is blocking everyone else.

We need this policy to declare the sweet point that we follow. Hopefully, as we improve our tests, we can tighten the two parameters (probability of failure and expected resolution time).

Reply:
Couple of points. Imho this policy in question is totally context depended. Some lanes are more critical than others, sometime a lane can fail only on specific Kubernetes version (or any dependency per se).  It also depend on why the lane fails. Sometimes it can be out of hands of developers, me remembering vfio bug we discovered that lead to the vgpu lane being unusable.

We are talking about blocking but are we really blocking anyone? Our release process last for ~3 months, not merging any PRs for 3 working days would not hurt any group or individual, imho. Of course in case of bug/security fix we are talking about different context and there maintainer can decide after throughout  considerations to override the context.

Rather than defining a strict policy of quarantining a whole lane I ask to think about process that could help me (as gate keeper but others as well) and contributors to guide through such a rare crisis.

Reply:
> Couple of points. Imho this policy in question is totally context depended. Some lanes are more critical than others, sometime a lane can fail only on specific Kubernetes version (or any dependency per se). It also depend on why the lane fails. Sometimes it can be out of hands of developers, me remembering vfio bug we discovered that lead to the vgpu lane being unusable.

Yes I agree with you - the context of the failures is very important. SIG CI would take this context into consideration and discuss with the responsible SIG before taking any action. This doc update is to act as a guideline for SIG CI.


> 
> We are talking about blocking but are we really blocking anyone? Our release process last for ~3 months, not merging any PRs for 3 working days would not hurt any group or individual, imho. Of course in case of bug/security fix we are talking about different context and there maintainer can decide after throughout considerations to override the context.

3 days without merging could lead to say ~30 PRs in the queue - this can lead to its own problems with CI load and unnecessary retests but also a large number of different combinations of batches runs being created. Its just wasteful. 

Also it can be frustrating for the other SIGs trying to deliver features unrelated to the flake. 

> 
> Rather than defining a strict policy of quarantining a whole lane I ask to think about process that could help me (as gate keeper but others as well) and contributors to guide through such a rare crisis.

Reply:
@xpivarc I agree that the context is important; I tend to agree with you that 3 days may be two short, particularly if it is a weekend. However where would *you* draw the line?

Let us assume that the windows lane is suddenly failing 90% of the times, making PRs merge after ~10 retries and many hours. For how long would you say we should accept this, if it's more than three days - a week? A month?

Every day a PR is not merged due to a reason unrelated to itself (say, a faulty windows lane), we lose the attention of the PR contributor, cause frustration and of course economic cost. At some point, SIG CI should put its foot down and force the owners of the faulty lane to have it fixed ASAP.

Reply:
So I would suggest to have a week as period of time where we acknowledge the problem, relevant sig is informed and is making it priority to fix the issues. After the week I would opt-in to quarantine. I am wondering what and if anything can be done to ensure quality while the lane is quarantined.

What do we think?

Reply:
> So I would suggest to have a week as period of time where we acknowledge the problem, relevant sig is informed and is making it priority to fix the issues. After the week I would opt-in to quarantine. I am wondering what and if anything can be done to ensure quality while the lane is quarantined.
> 
> What do we think?

OK a week period seems fair. This will allow us to use the test lane badges in ci-health as a metric on whether to quarantine or not. I will make updates to capture this.